File: app/src/main/java/com/akilimo/mobile/views/fragments/PrivacyStatementFragment.java
package com.akilimo.mobile.views.fragments;


import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatCheckBox;
import androidx.fragment.app.Fragment;

import com.akilimo.mobile.R;
import com.akilimo.mobile.databinding.FragmentPrivacyStatementBinding;
import com.akilimo.mobile.inherit.BaseStepFragment;
import com.akilimo.mobile.utils.SessionManager;
import com.stepstone.stepper.VerificationError;

/**
 * A simple {@link Fragment} subclass.
 * https://app-privacy-policy-generator.firebaseapp.com/#
 */
public class PrivacyStatementFragment extends BaseStepFragment {

    FragmentPrivacyStatementBinding binding;
    private boolean policyAccepted;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        this.context = context;
    }

    public PrivacyStatementFragment() {
        // Required empty public constructor
    }

    public static PrivacyStatementFragment newInstance() {
        return new PrivacyStatementFragment();
    }


    @Override
    protected View loadFragmentLayout(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentPrivacyStatementBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }


    @SuppressLint("SetJavaScriptEnabled")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (sessionManager != null) {
            sessionManager = new SessionManager(context);
        }

        WebView.setWebContentsDebuggingEnabled(false);
        binding.webView.setWebChromeClient(new WebChromeClient());
        binding.webView.setScrollContainer(true);
        binding.webView.setVerticalScrollBarEnabled(false);
        binding.webView.setScrollBarStyle(View.SCROLLBARS_INSIDE_OVERLAY);
        binding.webView.setHorizontalScrollBarEnabled(false);
        binding.webView.setWebChromeClient(new WebChromeClient());

        WebSettings webSettings = binding.webView.getSettings();

        webSettings.setJavaScriptEnabled(true);
        webSettings.setDomStorageEnabled(true);
        webSettings.setBuiltInZoomControls(true);
        webSettings.setJavaScriptEnabled(true);
        webSettings.setSupportZoom(true);
        webSettings.setDisplayZoomControls(false);
        webSettings.setLoadWithOverviewMode(true);
        webSettings.setUseWideViewPort(true);
        webSettings.setGeolocationEnabled(true);


        binding.chkAgreeToTerms.setOnCheckedChangeListener((compoundButton, checked) -> policyAccepted = checked);
    }

    @Override
    public void onSelected() {
        String termsLink = sessionManager.getTermsLink();
        binding.webView.loadUrl(termsLink);
    }


    @Nullable
    @Override
    public VerificationError verifyStep() {
        sessionManager.setTermsAccepted(policyAccepted);
        if (policyAccepted) {
            //save to session and skip in future startup
            return null;
        }
        return new VerificationError(getString(R.string.lbl_accept_terms_prompt));
    }


    @Override
    public void onError(@NonNull VerificationError error) {

    }
}

File: app/src/main/java/com/akilimo/mobile/views/fragments/PrivacyStatementFragment.kt
package com.akilimo.mobile.views.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.lifecycle.lifecycleScope
import com.akilimo.mobile.R
import com.akilimo.mobile.databinding.FragmentPrivacyStatementBinding
import com.akilimo.mobile.inherit.BindBaseStepFragment
import com.stepstone.stepper.VerificationError
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class PrivacyStatementFragment : BindBaseStepFragment<FragmentPrivacyStatementBinding>() {

    companion object {
        const val SCROLL_DELAY = 300L
        fun newInstance(): PrivacyStatementFragment = PrivacyStatementFragment()
    }

    private var scrollCheckJob: Job? = null

    override fun inflateBinding(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ) = FragmentPrivacyStatementBinding.inflate(inflater, container, false)

    override fun onBindingReady(savedInstanceState: Bundle?) {
        WebView.setWebContentsDebuggingEnabled(false)

        binding.privacyStatementWebView.run {
            webViewClient = object : WebViewClient() {
                override fun onPageFinished(view: WebView?, url: String?) {
                    super.onPageFinished(view, url)
                    startScrollMonitoring()
                }

                override fun shouldOverrideUrlLoading(
                    view: WebView?,
                    request: WebResourceRequest?
                ): Boolean = false
            }

            isScrollContainer = true
            isVerticalScrollBarEnabled = false
            isHorizontalScrollBarEnabled = false
            scrollBarStyle = View.SCROLLBARS_INSIDE_OVERLAY

            settings.apply {
                javaScriptEnabled = true
                domStorageEnabled = true
                builtInZoomControls = true
                displayZoomControls = false
                loadWithOverviewMode = true
                useWideViewPort = true
                setSupportZoom(true)
                setGeolocationEnabled(false)
            }

            loadUrl(preferenceManager.privacyPolicyLink)
        }
    }

    private fun startScrollMonitoring() {
        scrollCheckJob?.cancel() // cancel any previous job

        scrollCheckJob = viewLifecycleOwner.lifecycleScope.launch {
            while (!preferenceManager.privacyPolicyRead) {
                val webView = binding.privacyStatementWebView
                if (webView.isScrolledToBottom()) {
                    preferenceManager.privacyPolicyRead = true
                    break
                }
                delay(SCROLL_DELAY) // check every 300ms
            }
        }
    }

    override fun setupObservers() {
        TODO("Not yet implemented")
    }

    override fun verifyStep(): VerificationError? {
        return if (preferenceManager.privacyPolicyRead) {
            null
        } else {
            VerificationError(getString(R.string.lbl_read_full_policy_prompt))
        }
    }

    override fun onDestroyView() {
        scrollCheckJob?.cancel()
        scrollCheckJob = null
        super.onDestroyView()
    }
}
